src/utils.ts

```diff
@@ -227,3 +229,78 @@ export function arrify<T>(val: T | T[]) {
 
   return Array.isArray(val) ? val : [val];
 }
+
+export function ensureProgram(instance: TSInstance) {
+  if (instance && instance.watchHost) {
+    if (instance.hasUnaccountedModifiedFiles) {
+      if (instance.changedFilesList) {
+        instance.watchHost.updateRootFileNames();
+      }
+      if (instance.watchOfFilesAndCompilerOptions) {
+        instance.program = instance.watchOfFilesAndCompilerOptions
+          .getProgram()
+          .getProgram();
+      }
+      instance.hasUnaccountedModifiedFiles = false;
+    }
+    return instance.program;
+  }
+  if (instance.languageService) {
+    return instance.languageService.getProgram();
+  }
+  return instance.program;
+}
+
+export function supportsProjectReferences(instance: TSInstance) {
+  const program = ensureProgram(instance);
+  return program && !!program.getProjectReferences;
+}
+
+export function isUsingProjectReferences(instance: TSInstance) {
+  if (supportsProjectReferences(instance)) {
+    const program = ensureProgram(instance);
+    return program && program.getProjectReferences();
+  }
+  return false;
+}
+
+export function getProjectReferenceForFile(
+  filePath: string,
+  instance: TSInstance
+) {
+  if (isUsingProjectReferences(instance)) {
+    const program = ensureProgram(instance);
+    return (
+      program &&
+      program
+        .getProjectReferences()!
+        .find(ref => ref!.commandLine.fileNames.indexOf(filePath) > -1)
```

Comment:
do you handle missing referenced projects somewhere else or how can you be sure `ref` is always defined?
`ref` will be `undefined` if a referenced project cannot be resolved, i.e. missing folder or missing tsconfig.json.

Reply:
It will be a holey array? I didn't realize that; it feels a bit unintuitive, but I guess the types agree. Let me do something here. Thanks for the feedback!